hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -1125,11 +1128,12 @@ private void addCollection(PersistentCollection<?> coll, CollectionEntry entry,
 		}
 	}
 
-	private IdentityMap<PersistentCollection<?>, CollectionEntry> getOrInitializeCollectionEntries() {
+	private void putCollectionEntry(PersistentCollection<?> collection, CollectionEntry entry) {
 		if ( this.collectionEntries == null ) {
-			this.collectionEntries = IdentityMap.instantiateSequenced( INIT_COLL_SIZE );
+			this.collectionEntries = new InstanceIdentityMap<>();
 		}
-		return this.collectionEntries;
+		collection.$$_hibernate_setInstanceId( nextCollectionInstanceId() );
```

Comment:
```suggestion
		assert collection.$$_hibernate_getInstanceId() == -1;
		collection.$$_hibernate_setInstanceId( nextCollectionInstanceId() );
```

Reply:
The collection might have been associated to a different persistence context, we don't reset instance ids.

Reply:
Maybe I'm missing something, but didn't we discuss that we could unset the instance id in `unsetSession()` when changing the order of operations that call this method right now?
I'm not saying this check is super critical to have, just that it might be a nice way to capture potential bugs in our code.

Reply:
The resetting happens in `org.hibernate.engine.internal.EntityEntryContext#removeEntityEntry` and there are currently 2 itable callsites there, I suggest we introduce a new method that handles "reset" which manages previous/next entries and the instance id, which will even improve perf further :)

Reply:
A single method for previous/next is a great improvement idea! Why not do that as a separate improvement? 

I would not stress too much on resetting ids though, as we've already discussed there might still be cases which we need to handle (with the `==` check), and it will require an additional check on the id when accessing the collection, so I don't see that much additional value by doing that.

Reply:
>A single method for previous/next is a great improvement idea! Why not do that as a separate improvement?

Sounds good to me.

>I would not stress too much on resetting ids though, as we've already discussed there might still be cases which we need to handle (with the == check), and it will require an additional check on the id when accessing the collection, so I don't see that much additional value by doing that.

Like I wrote before, it's not super critical to have this, it would IMO just be nice because it can be a nice safety belt for us. Since we'd only check the id in assertions, regular code would not be impacted.

Reply:
https://hibernate.atlassian.net/browse/HHH-19070